enhancements/oc/olm-mirroring.md

```diff
@@ -0,0 +1,121 @@
+
+$ docker pull disconnected-registry:5000/openshift/community-operators-catalog:4.2.1
+
+$ opm registry images --from=disconnected-registry:5000/openshift/community-operators-catalog:4.2.1 --to=disconnected-registry:5000/community-operators --manifests=./mirror-manifests | xargs oc image mirror
+
+$ oc apply -f ./mirror-manifests
+```
+
+This requires both `oc` and `opm` to be installed and available on bastion hosts. We would like to distribute one tool for creating disconnected tools, not two.
+
+### Goals
+* Define `oc` commands which can build and mirror all images required for mirroring to a disconnected envorinment
+
+### Non-Goals
+* Provide an api for fine-grained control over which operators / operands are mirrored. 
+
+## Proposal
+
+The basic approach will be to:
+
+- Generate catalog container images that can replace the appregistry repositories
+- For each operator within the catalog image, mirror it to the disconnected environment
+
+### Generate and mirror a catalog
+
+`oc` will be extended with a command (equivalent to the `opm` command)
+
+```sh
+$ oc adm catalog build --appregistry-namespace=community-operators --to quay.io/ecordell/community-operators-catalog:4.2.1 
+```
+
+`oc adm catalog build` will:
+
+- Use appregistry protocol to retrieve all of the operators in a namespace of Quay.io specified by `appregistry-namespace`
+- Load all of the downloaded operators into a versioned `operator-registry` sqlite artifact
+- Build a runnable operator-registry image by appending the database to the `operator-registry` base image using the same machinary as `oc image append`
+- Mirror that image to the target defined by `--to`, using the same machinary as `oc image append`
+
+It will have the following flags:
+
+- `--from=ref` - the base image to add the built operator database into. Defaults to the operator-registry image shipped with the version of OpenShift that `oc` came from.
+- `--to=ref` - the location that the image will be mirrored.
+- `--save=path` - instead of mirroring, this will output a tarball containing the image
+- `--auth-token=string` - the auth token ([instructions](https://github.com/operator-framework/operator-courier#authentication)) for authenticating with appregistry.
+- `--appregistry-endpoint=url` - the CNR endpoint to authenticate against. Defaults to `"https://quay.io/cnr"`, the endpoint used by OpenShift 4.1-4.3.
+- `--appregistry-namespace=string` - the namespace in appregistry to mirror. Each repository within the namespace represents one operator. 
+- `--db-only` - if true, and `--save` is specified, the operator database will be saved in the directory noted by `--save`.
+
+This command generates and mirrors in one step, because we do not assume that there is any registry available aside from the target disconnected registry. `--save` can be used to save the image locally without a registry available.
```

Comment:
I'm a bit confused by how `--save` is being used.

Reply:
It's the same as `docker save`, you get a tarball with the image contents / manifests / etc

Reply:
I'm going to remove this and the corresponding load; I think we can add them later if needed.